Add maven pom for build

geomacy opened the pull request:
This adds a maven pom.xml so that the project can be conveniently incorporated into an overall Maven build.  It assumes the toolchain exists, i.e. you have Go and Godep already set up.

You can invoke Maven as any of 
- `mvn clean install`                                     build for all supported platforms
- `mvn -Dtarget=native clean install`                     build for the current platform
- `mvn -Dtarget=cross -Dos=OS -Darch=ARCH clean install`  build for platform with operating system OS and architecture ARCH

where the supported OSes are any of darwin freebsd linux netbsd openbsd windows, and the architectures are either 386 or amd64.

The maven artefact is a zip file containing all the built binaries.

geomacy committed:
Add a Maven pom to support build from parent Brooklyn repo.

geomacy committed:
Update README and build script.

README updated with latest details.

Build script updated to run from its own directory, brooklyn-cli/release.

geomacy committed:
Add assembly to create the maven artifact.

geomacy committed:
Add -ldflags "-s" to reduce binary sizes a little.

m4rkmckenna commented on src/main/assembly/bin.xml:
```diff
@@ -0,0 +1,24 @@
+          xsi:schemaLocation="http://maven.apache.org/plugins/maven-assembly-plugin/assembly/1.1.3 http://maven.apache.org/xsd/assembly-1.1.3.xsd">
+    <id>bin</id>
+    <formats>
+        <format>zip</format>
```
Consider adding all artifacts to maven with different classifiers

lloyddave commented on README.md:
```diff
@@ -1,19 +1,49 @@
+## Toolchain
+
+You will need the following tools to build the CLI:
+- Go (min version 1.5.1), with full cross-compiler support (see https://golang.org/doc/install/source).
```
For this Toolchain section, maybe better to just provide a high level link (eg https://golang.org/doc/) rather than jumping to install info.  So keep it descriptive of the tools.

lloyddave commented on README.md:
```diff
@@ -1,19 +1,49 @@
+
+
+## Build Pre-Requisites
+
```
Info on installing tools here:
- Install Go (`https://golang.org/dl/`)  
  On a Mac, you use for example Homebrew (`brew install go --with-cc-all`)
- Set GOPATH environment variable
- Download and install Godep  
  `go get github.com/tools/godep`
- Install Brooklyn CLI and dependencies  
  `go get -d o get github.com/brooklyncentral/brooklyn-cli/br`

lloyddave commented on README.md:
```diff
@@ -1,19 +1,49 @@
+- Ensure your [$GOPATH](http://golang.org/cmd/go/#hdr-GOPATH_environment_variable) is set correctly 
+  to a suitable location for your Go code.
+- git clone the CLI code into $GOPATH/src/github.com/brooklyncentral/brooklyn-cli
+
```
Add a section for 'Compiling the code with Go'.  This section should be using godep for build/install so that it uses the 'cached' dependencies.

```
cd $GOPATH/src/github.com/brooklyncentral/brooklyn-cli/br
godep go install
```

lloyddave commented on README.md:
```diff
@@ -1,19 +1,49 @@
+- git clone the CLI code into $GOPATH/src/github.com/brooklyncentral/brooklyn-cli
+
+## Compiling the code with Go for development purposes
+
```
This section should maybe detail on using the 'external' dependencies rather than those cached with godep.  Would need to mention that if a compile error occurs then to try using the godep cached dependencies; also potential for using `go clean -i` and/or `godep go clean -i` to remove old object/archive files.

lloyddave commented on release/build.sh:
```diff
@@ -175,16 +171,15 @@ elif [ -z "$all" ]; then
 		for i in $os; do
-			printf "$i/$j "
-			GOOS="$i" GOARCH="$j" $GODEP $GOBIN build -o "${dir}/${BRNAME}${label}${timestamp}.$i.$j"
+			printf "    $i/$j \n"
```
Rather than using `printf "...\n"` it would be better to just use `echo "..."`; then remove line 184.

lloyddave replied:
Rather than using `printf "...\n"` it would be better to just use `echo "..."`; then remove line 184.

geomacy committed:
Update README and build script per review comments.

At this stage I haven't added separate maven artifacts for separate platforms, will look into that at a later stage.

geomacy committed:
Add a line to README to say what the tool is.

lloyddave commented:
LGTM.  Merging.
